controllers/topology/conditions.go

```diff
@@ -0,0 +1,119 @@
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/scope"
+	"sigs.k8s.io/cluster-api/util/conditions"
+)
+
+func (r *ClusterReconciler) reconcileConditions(s *scope.Scope, cluster *clusterv1.Cluster, reconcileErr error) error {
+	return r.reconcileTopologyReconciledCondition(s, cluster, reconcileErr)
+}
+
+// reconcileTopologyReconciledCondition sets the TopologyReconciled condition on the cluster.
+// The TopologyReconciled condition is considered true is spec of all the objects associated with the
+// cluster is in sync with the topology defined in the cluster.
+// The condition is false under the following conditions:
+// - An error occurred during the reonciled process of the cluster topology.
+// - The cluster upgrade has not yet propogated to all the components of the cluster.
+//   - For a managed topology cluster the version upgrade is propagated one component at a time.
+//     In such a case since some of the component's sopec could be adrift from the topology the
+//     topology cannot be considered fully reconciled.
+func (r *ClusterReconciler) reconcileTopologyReconciledCondition(s *scope.Scope, cluster *clusterv1.Cluster, reconcileErr error) error {
+	// If an erorr occurred during reconciliation set the TopologyReconciled condition to false.
+	// Add the error message from the reoncile function to the message of the condition.
+	if reconcileErr != nil {
+		conditions.Set(
+			cluster,
+			conditions.FalseCondition(
+				clusterv1.TopologyReconciledCondition,
+				clusterv1.TopologyReconciledErroredReason,
+				clusterv1.ConditionSeverityError,
+				"An error occurred while reconciling the cluster topology. Error: %s", reconcileErr,
+			),
+		)
+		return nil
+	}
+
+	// If either the Control Plane or any of the MachineDeployments are still pending to pick up the new version (generally
+	// happened when upgrading the cluster) then the topology is not considered as fully reconciled.
+	if !s.UpgradeTracker.ControlPlane.IsAtTopologyVersion || s.UpgradeTracker.MachineDeployments.HasPendingUpdates() {
+		var msg string
+		var reason string
+		if !s.UpgradeTracker.ControlPlane.IsAtTopologyVersion {
+			msg = fmt.Sprintf("Upgrade to %s on hold. ", s.Blueprint.Topology.Version)
+			reason = clusterv1.TopologyReconciledControlPlaneUpdatePendingReason
+		} else {
+			msg = fmt.Sprintf("Update to MachineDeployment(s) %s on hold. ",
+				strings.Join(s.UpgradeTracker.MachineDeployments.PendingUpdateNames(), ", "),
+			)
+			reason = clusterv1.TopologyReconciledMachineDeploymentsUpdatePendingReason
+		}
+
+		switch {
+		case s.UpgradeTracker.ControlPlane.IsProvisioning:
+			msg += "Control Plane is provisioning"
+
+		case s.UpgradeTracker.ControlPlane.IsUpgrading:
+			cpVersion, err := contract.ControlPlane().Version().Get(s.Current.ControlPlane.Object)
+			if err != nil {
+				return errors.Wrap(err, "failed to get control plane spec version")
+			}
+			msg += fmt.Sprintf("Control plane is upgrading to %s", *cpVersion)
```

Comment:
```suggestion
			msg += fmt.Sprintf("Control plane is in the procees of completing a previous upgrade to %s", *cpVersion)
```

Reply:
Similar to https://github.com/kubernetes-sigs/cluster-api/pull/5732#discussion_r758635456 it probably would be nice to know what version the control plane is upgrading to. In cases where multiple version upgrades were queued it this information will be useful. How about a message like `Control plane is in the process of completing a previous upgrade to version %s` ?

Reply:
if I'm not wrong It matches exactly my suggestion

Reply:
You are right. I misread the suggestion. 😅